sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -528,6 +528,40 @@ object SimplifyConditionals extends Rule[LogicalPlan] with PredicateHelper {
 }
 
 
+/**
+ * Push the foldable expression into (if / case) branches.
+ */
+object PushFoldableIntoBranches extends Rule[LogicalPlan] with PredicateHelper {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+    case q: LogicalPlan => q transformExpressionsUp {
+      case b @ BinaryExpression(i @ If(_, trueValue, falseValue), right) if i.deterministic &&
+          right.foldable && (trueValue.foldable || falseValue.foldable) =>
```

Comment:
Sorry if I missed something but can you explain why we need this foldable condition?

Shouldn't `bin_op((if cond: a else: b), c)` be same as `if cond: bin_op(a, c) else: bin_op(b, c)` as long as the results are deterministic?

Reply:
Ohh, okay but we just want to push down foldables only that can have benefits. gotya.